File: JavaScript/프로그래머스/2/42888. 오픈채팅방/오픈채팅방.js
function solution(record) {
    var answer = [];
    const action = []
    const nameObj = {};
    for (const str of record) {
        const tempArr = str.split(' ');
        if (tempArr[0][0] === 'E'){
            nameObj[tempArr[1]] = tempArr[2];
            action.push(['E', tempArr[1]]);
        } else if  (tempArr[0][0] === 'L'){
            action.push(['L', tempArr[1]]);
        } else {
            nameObj[tempArr[1]] = tempArr[2];
        }
    }
    for (const [act, id] of action){
        if (act === 'E'){
            answer.push(`${nameObj[id]}님이 들어왔습니다.`);
        } else {
            answer.push(`${nameObj[id]}님이 나갔습니다.`);
        }
    }
    return answer;
}

File: C++/프로그래머스/2/42888. 오픈채팅방/오픈채팅방.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <utility>
#include <iostream>

using namespace std;

vector<string> solution(vector<string> record) {
    vector<string> answer;
    unordered_map<string, string> mp;
    vector<pair<string, string>> course;
    for (auto mem: record){
        if (mem[0] == 'E'){
            int pos = mem.find(' ', 6);     
            mp[mem.substr(6 , pos - 6)] = mem.substr(pos + 1);
            course.push_back({"Enter", mem.substr(6 ,pos - 6)});
        } else if (mem[0] == 'L'){
            course.push_back({"Leave", mem.substr(6)});
        } else {
            int pos = mem.find(' ', 7);   
            mp[mem.substr(7, pos - 7)] = mem.substr(pos + 1);
        }
    }
    for (auto mem: course){
        string words = "";
        words += mp[mem.second];
        if (mem.first == "Enter"){
            words += "님이 들어왔습니다.";
        } else {
            words += "님이 나갔습니다.";
        }
        answer.push_back(words);
    }
    return answer;
}